src/ucp/core/ucp_worker.c

```diff
@@ -481,12 +481,27 @@ static unsigned ucp_worker_iface_err_handle_progress(void *arg)
             continue;
         }
 
-        ucs_trace("ep %p: discard uct_ep[%d]=%p", ucp_ep, lane,
-                  ucp_ep->uct_eps[lane]);
-        ucp_worker_discard_uct_ep(ucp_ep->worker, ucp_ep->uct_eps[lane],
-                                  UCT_FLUSH_FLAG_CANCEL,
-                                  ucp_ep_err_pending_purge,
-                                  UCS_STATUS_PTR(status));
+        if (lane == ucp_ep_get_cm_lane(ucp_ep)) {
+            /* Don't schedule UCT CM lane to be discarded, since UCP EP will be
+             * destroyed due to peer failure and ucp_cm_disconnect_cb() could be
+             * invoked on async thread after UCP EP is destroyed and before UCT
+             * CM EP is destroyed from discarding functionality. So, UCP EP will
+             * passed as a corrupted argument to ucp_cm_disconnect_cb() */
+            ucs_trace("ep %p: destroy uct_ep[%d]=%p", ucp_ep, lane,
+                      ucp_ep->uct_eps[lane]);
+            uct_ep_pending_purge(ucp_ep->uct_eps[lane],
+                                 ucp_ep_err_pending_purge,
+                                 UCS_STATUS_PTR(status));
+            uct_ep_destroy(ucp_ep->uct_eps[lane]);
+        } else {
+            ucs_trace("ep %p: discard uct_ep[%d]=%p", ucp_ep, lane,
+                      ucp_ep->uct_eps[lane]);
+            ucp_worker_discard_uct_ep(ucp_ep->worker, ucp_ep->uct_eps[lane],
+                                      UCT_FLUSH_FLAG_CANCEL,
+                                      ucp_ep_err_pending_purge,
+                                      UCS_STATUS_PTR(status));
```

Comment:
can we change discard logic, so it will destroy ep immediately, if flush is completed?
since cm flush should complete immediately

Reply:
ok, done